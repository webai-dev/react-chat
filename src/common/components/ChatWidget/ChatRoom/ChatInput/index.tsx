import React, {useState, useContext, useRef, useEffect} from "react";
import { Container } from "@material-ui/core";
import { makeStyles } from "@material-ui/styles";
import SendIcon from "@material-ui/icons/Send";
import chat from "../../../../services/chat";
import { ChatRoom } from "../../../../../types/types";
import { MeContext } from "../../../../hooks/useMe";
import { CompanyContext } from "../../../../hooks/useCompanies";
import { ChatRoomsContext } from "../../../../hooks/useChatRooms";
import { RoomListsContext } from "../../../../hooks/useRoomLists";
import { v4 as uuidv4 } from "uuid";
import InputEmoji from "react-input-emoji";
import { NewMessage } from "../../../../../types";

// Global styles
export const useStyles = makeStyles({
  icon: {
    width: "31px",
    height: "31px",
    padding: "0 10px",
    cursor: "pointer",
  },
  input: {
    outline: "none",
    height: "39px",
    width: "100%",
    padding: "0 10px",
    border: "1px solid #D6D6D6",
  },
  root: {
    display: "flex",
    flexDirection: "row",
    alignItems: "center",
    justifyContent: "space-between",
    width: "100%",
    height: "55px",
    padding: "5px 15px",

    color: "#3870A0",
    bottom: 0,
    fontSize: "24px",
  },
});

interface ChatInputProps {
  chatRoom: ChatRoom | null | undefined;
  connectionStatus: number;
  setSendFailed: (flag: boolean) => void;
}

/**
 * ChatInput component
 * @param {ChatInputProps} props Properties
 * @description ChatInput component consists of input box and send button
 */
const ChatInput: React.FC<ChatInputProps> = ({
  chatRoom = null,
  connectionStatus = 1,
  setSendFailed,
}) => {
  // Default styling
  const classes = useStyles();
  const [message, setMessage] = useState<string>("");
  const [isSending, setIsSending] = useState<boolean>(false);
  const { sendMessage } = chat;
  const { me } = useContext(MeContext);
  const { selectedCompany } = useContext(CompanyContext);
  const { updateChatRooms } = useContext(ChatRoomsContext);
  const { companyRoomLists } = useContext(RoomListsContext);
  const inputRef = useRef() as React.MutableRefObject<HTMLInputElement>;

  const handleChange = (e: string) => {
    setMessage(e);
  };

  // useEffect(() => {
  //   console.log(connectionStatus);
  // }, [connectionStatus]);

  const handleSend = async () => {
    let data = {
      message: message,
      trackingID: uuidv4(),
      sentAt: new Date(),
    };
    setIsSending(true);
    setMessage("");
    if (chatRoom && me) {
      chatRoom.messages = [
        ...chatRoom.messages,
        {
          id: 0,
          message: message,
          messageType: 0,
          readBy: me.id,
          sentAt: data.sentAt.toString(),
          trackingID: data.trackingID,
          user: {
            id: me.id,
            firstName: me.firstName,
            lastName: me.lastName,
            isOnline: true,
            companyMembership: [],
          },
          new: true,
          check: "sent",
        },
      ];
      chatRoom.newMessageFromMe = true;

      updateChatRooms(chatRoom);
    }
    let updatedCompanyRooms = companyRoomLists.find(
      (company) => selectedCompany && company.companyId === selectedCompany.id
    );
    if (updatedCompanyRooms && updatedCompanyRooms.rooms) {
      let targetRoom = updatedCompanyRooms.rooms.find(
        (r) => chatRoom && r && r.id === chatRoom.roomId
      );
      if (targetRoom) {
        updatedCompanyRooms.rooms = [
          targetRoom,
          ...updatedCompanyRooms?.rooms.filter(
            (r) => chatRoom && r && r.id !== chatRoom.roomId
          ),
        ];
      }
    }
    if (connectionStatus === 1) {
      const response = await sendMessage(
        selectedCompany && selectedCompany.id,
        chatRoom?.roomId,
        data
      );
      if (!response && chatRoom?.messages) {
        chatRoom.messages = chatRoom?.messages.map((m) =>
          m.id === 0 ? { ...m, check: "failed" } : m
        );
      }
    } else {
      let response: NewMessage | null = null;
      if (!(connectionStatus === 1)) {
        response = await sendMessage(
            selectedCompany && selectedCompany.id,
            chatRoom?.roomId,
            data
        );

        if (response) {
          if (chatRoom?.messages) {
            chatRoom.messages = chatRoom?.messages.map((m) =>
                m.id === 0 ? { ...m, check: "received" } : m
            );
            updateChatRooms(chatRoom);
          }
        } else {
          let started = Date.now();
          // loop every 10 seconds
          let interval = setInterval(async function () {
            if (Date.now() - started > 46000) {
              clearInterval(interval);
              if (chatRoom?.messages) {
                chatRoom.messages = chatRoom?.messages.map((m) =>
                    m.id === 0 ? { ...m, check: "failed" } : m
                );
                setSendFailed(true);
              }
            }
            // Retry sending message for first 2 minutes or after 5 minutes
            else {
              // let response = connectionStatus === 1 ? 1 : 0;
              if (response) {
                if (chatRoom?.messages) {
                  chatRoom.messages = chatRoom?.messages.map((m) =>
                      m.id === 0 ? { ...m, check: "received" } : m
                  );
                  updateChatRooms(chatRoom);
                }
                setSendFailed(false);
                clearInterval(interval);
              } else {
                response = await sendMessage(
                    selectedCompany && selectedCompany.id,
                    chatRoom?.roomId,
                    data
                );
              }
            }
          }, 15000);
        }

      }
    }
    setIsSending(false);
  };

  useEffect(() => {
    if(isSending)
      inputRef.current.focus();
  }, [inputRef.current]);

  return (
    <>
      <Container data-testid="chat-input" className={classes.root}>
        <InputEmoji
          value={message}
          className={classes.input}
          onChange={handleChange}
          cleanOnEnter
          onEnter={() => {
            handleSend();
            // inputRef.current.autofocus = true;
          }}
          placeholder="Message"
          ref={inputRef}
        />

        <SendIcon onClick={handleSend} className={classes.icon} />
      </Container>
    </>
  );
};

export default ChatInput;
