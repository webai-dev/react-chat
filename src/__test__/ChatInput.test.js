import React from "react";
import { render } from "@testing-library/react";
import "@testing-library/jest-dom";
import ChatInput from "../common/components/ChatWidget/ChatRoom/ChatInput";
import renderer from "react-test-renderer";

// react-emoji package error in test
// it("renders without crashing", () => {
//   const { getByTestId } = render(<ChatInput />);
//   const chatInput = getByTestId("chat-input");
//   expect(chatInput).toBeTruthy();
// });

it("matches snapshot", () => {
  const tree = renderer.create(<ChatInput />).toJSON();
  expect(tree).toMatchSnapshot();
});
