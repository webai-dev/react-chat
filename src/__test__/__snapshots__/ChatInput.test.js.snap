// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<div
  className="MuiContainer-root makeStyles-root-3 MuiContainer-maxWidthLg"
  data-testid="chat-input"
>
  <div
    className="react-emoji"
  >
    <div
      className="react-emoji-picker--container"
    >
      <div
        className="react-emoji-picker--wrapper"
      >
        <div
          className="react-emoji-picker"
        />
      </div>
    </div>
    <div
      className="react-input-emoji--container"
      style={
        Object {
          "borderColor": "#EAEAEA",
          "borderRadius": 21,
          "fontFamily": "sans-serif",
          "fontSize": 15,
        }
      }
    >
      <div
        className="react-input-emoji--wrapper"
        onClick={[Function]}
      >
        <div
          className="react-input-emoji--placeholder"
        >
          Message
        </div>
        <div
          className="react-input-emoji--input"
          contentEditable={true}
          onBlur={[Function]}
          onKeyDown={[Function]}
          tabIndex={0}
        />
      </div>
    </div>
    <button
      className="react-input-emoji--button"
      onClick={[Function]}
    >
      <svg
        height="24"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M12 0C5.373 0 0 5.373 0 12s5.373 12 12 12 12-5.373 12-12S18.627 0 12 0m0 22C6.486 22 2 17.514 2 12S6.486 2 12 2s10 4.486 10 10-4.486 10-10 10"
        />
        <path
          d="M8 7a2 2 0 1 0-.001 3.999A2 2 0 0 0 8 7M16 7a2 2 0 1 0-.001 3.999A2 2 0 0 0 16 7M15.232 15c-.693 1.195-1.87 2-3.349 2-1.477 0-2.655-.805-3.347-2H15m3-2H6a6 6 0 1 0 12 0"
        />
      </svg>
    </button>
  </div>
  <svg
    aria-hidden={true}
    className="MuiSvgIcon-root makeStyles-icon-1"
    focusable="false"
    onClick={[Function]}
    viewBox="0 0 24 24"
  >
    <path
      d="M2.01 21L23 12 2.01 3 2 10l15 2-15 2z"
    />
  </svg>
</div>
`;
